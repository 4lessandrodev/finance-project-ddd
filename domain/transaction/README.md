# Transaction - Aggregate

```json
{
  "userId": "uuid",
  "totalValue": 100, //Dinamic calculations
  "reasonId": "uuid",
  "paymentDate": "2021-01-01 10:00:00", // Value Object
  "transaction-type": ["Entrada", "Saida"], // Value Object
  "status": ["Pendente", "Conclu√≠do"], // Value Object
  "note": "valid_description", // Value Object
  "attachment": "url", // Value Object
  "calculations": [
    {
      "budgetBoxId": "uuid",
      "Value": 50
    },
    {
      "budgetBoxId": "uuid",
      "Value": 50
    }
  ]
}
```

- transaction-type: Value Object (enum) - Ok
- status: Value Object (enum) - Ok
- note: Value Object
- attachment: Value Object
- calculations: Value Object
- paymentDate: Value Object - Ok
