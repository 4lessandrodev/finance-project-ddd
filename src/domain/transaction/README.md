# Transaction - Aggregate

```json
{
  "user-id": "uuid",
  "total-value": 100,
  "reasonId": "uuid",
  "payment-date": "2021-01-01 10:00:00",
  "transaction-type": "Entrada" | "Saida",
  "status": "Pendente" | "Conclu√≠do",
  "note": "valid_description",
  "attachment": "url",
  "transaction-calculations": [
    {
      "budgetbox-id": "uuid",
      "Value": 50
    },
    {
      "budgetbox-id": "uuid",
      "Value": 50
    }
  ]
}
```

- transaction-type: Value Object (enum) - Ok
- status: Value Object (enum) - Ok
- note: Value Object - Ok
- attachment-path: Value Object - Ok
- transaction-calculations: Value Object - Ok
- payment-date: Value Object - Ok
