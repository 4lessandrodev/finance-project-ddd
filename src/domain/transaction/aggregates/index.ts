export * from './transaction/transaction.aggregate';
