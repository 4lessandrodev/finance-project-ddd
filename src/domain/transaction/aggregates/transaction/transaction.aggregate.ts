import {
  AggregateRoot,
  DateValueObject,
  ReasonIdValueObject,
  Result,
  UniqueEntityID,
} from '@shared/index';
import {
  AttachmentPathValueObject,
  TransactionCalculationValueObject,
  TransactionNoteValueObject,
  TransactionStatusValueObject,
  TransactionTypeValueObject,
  UserIdValueObject,
} from '@domain/index';

export interface TransactionAggregateProps {
  userId: UserIdValueObject;
  reasonId: ReasonIdValueObject;
  paymentDate: DateValueObject;
  transactionType: TransactionTypeValueObject;
  status: TransactionStatusValueObject;
  transactionCalculations: TransactionCalculationValueObject[];
  note?: TransactionNoteValueObject;
  attachment?: AttachmentPathValueObject;
}

export class TransactionAggregate extends AggregateRoot<TransactionAggregateProps> {
  private _totalValue: number;
  private constructor(
    props: TransactionAggregateProps,
    total: number,
    id?: UniqueEntityID,
  ) {
    super(props, id);
    this._totalValue = total;
  }

  get totalValue(): number {
    return this._totalValue;
  }

  get userId(): UserIdValueObject {
    return this.props.userId;
  }

  get reasonId(): ReasonIdValueObject {
    return this.props.reasonId;
  }

  get paymentDate(): DateValueObject {
    return this.props.paymentDate;
  }

  get transactionType(): TransactionTypeValueObject {
    return this.props.transactionType;
  }

  get status(): TransactionStatusValueObject {
    return this.props.status;
  }

  get transactionCalculations(): TransactionCalculationValueObject[] {
    return this.props.transactionCalculations;
  }

  get note(): TransactionNoteValueObject | null {
    return this.props.note ?? null;
  }

  get attachment(): AttachmentPathValueObject | null {
    return this.props.attachment ?? null;
  }

  public static create(
    props: TransactionAggregateProps,
    id?: UniqueEntityID,
  ): Result<TransactionAggregate> {
    /**
     * total is calculated dinamicaly. Its the sum of calculation values
     */
    const total = props.transactionCalculations.reduce(
      (total, calc) => calc.calculation.value + total,
      0,
    );

    return Result.ok<TransactionAggregate>(
      new TransactionAggregate(props, total, id),
    );
  }
}
