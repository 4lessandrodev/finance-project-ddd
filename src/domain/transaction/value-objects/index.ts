export * from './attachment-path/attachment-path.value-object';
export * from './transaction-calculations/transaction-calculations.value-object';
export * from './transaction-note/transaction-note.value-object';
export * from './transaction-status/transaction-status.value-object';
export * from './transaction-type/transaction-type.value-object';
