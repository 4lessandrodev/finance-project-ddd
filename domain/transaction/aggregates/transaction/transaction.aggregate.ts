import {
  AggregateRoot,
  DateValueObject,
  ReasonIdValueObject,
  Result,
  UniqueEntityID,
} from '../../../shared';
import { UserIdValueObject } from '../../../user/value-objects';
import {
  AttachmentPathValueObject,
  TransactionCalculationValueObject,
  TransactionNoteValueObject,
  TransactionStatusValueObject,
  TransactionTypeValueObject,
} from '../../value-objects';

export interface TransactionAggregateProps {
  userId: UserIdValueObject;
  reasonId: ReasonIdValueObject;
  paymentDate: DateValueObject;
  transactionType: TransactionTypeValueObject;
  status: TransactionStatusValueObject;
  transactionCalculations: TransactionCalculationValueObject[];
  note?: TransactionNoteValueObject;
  attachment?: AttachmentPathValueObject;
}

export class TransactionAggregate extends AggregateRoot<TransactionAggregateProps> {
  private _totalValue: number;
  private constructor(props: TransactionAggregateProps, id?: UniqueEntityID) {
    super(props, id);
    this._totalValue = 0;
  }

  get totalValue(): number {
    return this._totalValue;
  }

  get userId(): UserIdValueObject {
    return this.props.userId;
  }

  get reasonId(): ReasonIdValueObject {
    return this.props.reasonId;
  }

  get paymentDate(): DateValueObject {
    return this.props.paymentDate;
  }

  get transactionType(): TransactionTypeValueObject {
    return this.props.transactionType;
  }

  get status(): TransactionStatusValueObject {
    return this.props.status;
  }

  get transactionCalculations(): TransactionCalculationValueObject[] {
    return this.props.transactionCalculations;
  }

  get note(): TransactionNoteValueObject | null {
    return this.props.note ?? null;
  }

  get attachment(): AttachmentPathValueObject | null {
    return this.props.attachment ?? null;
  }

  private calculateTotal(): number {
    const total = this.props.transactionCalculations.reduce(
      (total, calc) => calc.calculation.value + total,
      0,
    );
    this._totalValue = total;
    return total;
  }

  public static create(
    props: TransactionAggregateProps,
    id?: UniqueEntityID,
  ): Result<TransactionAggregate> {
    const transaction = new TransactionAggregate(props, id);
    transaction.calculateTotal();
    return Result.ok<TransactionAggregate>(transaction);
  }
}
