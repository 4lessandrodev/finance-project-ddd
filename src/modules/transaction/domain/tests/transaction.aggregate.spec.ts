import { TransactionMock } from './mock/transaction.mock';

describe('transaction.aggregate', () => {

	const transactionMock = new TransactionMock();

	it('should create a valid transaction', () => {
		
		const transaction = transactionMock.domain();

		expect(transaction.isSuccess).toBe(true);
		expect(transaction.getResult().note).toBeNull();
		expect(transaction.getResult().attachment).toBeNull();
	});

	it('should create a valid transaction with updated total', () => {
		
		const transaction = transactionMock.domain(
			{
				id: 'valid_id',
				transactionCalculations: [
					{
						budgetBoxName: 'valid_name',
						budgetBoxId: 'valid_id',
						currency: {
							currency: 'BRL',
							value: 170
						}
					},
					{
						budgetBoxName: 'valid_name',
						budgetBoxId: 'valid_id',
						currency: {
							currency: 'BRL',
							value: 30
						}
					}
				]
			}
		);
			
		expect(transaction.isSuccess).toBe(true);
		expect(transaction.getResult().totalValue.value).toBe(200);
	});

	it('should create a valid transaction with provided id ', () => {

		const transaction = transactionMock.domain({
			id: 'valid_id',
			attachment: 'https://aws.s3.com/bucket-askjdas656/file.pdf',
			note: 'valid_description'
		});

		expect(transaction.getResult().toObject()).toMatchSnapshot();
		expect(transaction.isSuccess).toBe(true);
		expect(transaction.getResult().id.toValue()).toBe('valid_id');
		expect(transaction.getResult().note?.value).toBe('valid_description');
		expect(transaction.getResult().attachment?.value).toBe('https://aws.s3.com/bucket-askjdas656/file.pdf');
	});

	it('should apply calculation with success', () => {

		const transaction = transactionMock.domain({
			id: 'valid_id',
			attachment: 'https://aws.s3.com/bucket-askjdas656/file.pdf',
			note: 'valid_description'
		}).getResult();

		
		const calculation = transaction.transactionCalculations[0];
		expect(transaction.transactionCalculations).toHaveLength(1);
		transaction.applyCalculation([calculation, calculation]);
		expect(transaction.transactionCalculations).toHaveLength(2);
	});
});
