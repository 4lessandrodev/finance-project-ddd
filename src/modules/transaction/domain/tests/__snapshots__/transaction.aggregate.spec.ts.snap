// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transaction.aggregate should create a valid transaction with provided id  1`] = `
Object {
  "attachment": "https://aws.s3.com/bucket-askjdas656/file.pdf",
  "createdAt": 2022-01-01T00:00:00.000Z,
  "deletedAt": undefined,
  "id": "valid_id",
  "isDeleted": false,
  "note": "valid_description",
  "paymentDate": 2022-01-01T00:00:00.000Z,
  "reason": "valid_reason",
  "status": "CONCLUIDO",
  "totalValue": Object {
    "currency": "BRL",
    "value": 1000,
  },
  "transactionCalculations": Array [
    Object {
      "budgetBoxId": "valid_budget_box_id",
      "budgetBoxName": "valid_name",
      "currency": Object {
        "currency": "BRL",
        "value": 1000,
      },
    },
  ],
  "transactionType": "ENTRADA",
  "updatedAt": 2022-01-01T00:00:00.000Z,
  "userId": "valid_user_id",
}
`;
